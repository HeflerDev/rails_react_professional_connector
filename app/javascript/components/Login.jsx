import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { connect } from 'react-redux';
import { loginUser } from '../actions/index';

const select = (dispatch) => ({
  loginUser: (user) => dispatch(loginUser(user)),
});

const mapStateToProps = (state) => ({
  isLoggedIn: state.userReducer.isLoggedIn,
});

const ConnectedLogin = ({ loginUser, isLoggedIn, history }) => {
  const [userData, setUserData] = useState({
    username: '',
    password: '',
    error: null,
  });

  useEffect(() => (isLoggedIn ? history.push('/') : null));

  const handleChange = (e) => {
    const { name, value } = e.target;
    setUserData({ ...userData, [name]: value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const { username, password } = userData;
    const user = {
      username,
      password,
    };
    axios.post('/login', { user }, { withCredentials: true })
      .then((res) => {
        if (res.data.logged_in) {
          loginUser(res.data.user);
          history.push('/');
        } else {
          throw new Error();
        }
      })
      .catch((err) => { setUserData({ ...userData, error: [err] }); });
  };

  const { username, password } = userData;

  return (
    <form onSubmit={handleSubmit} className="stack center signup-form">
      <label htmlFor="username" className="board center">
        <div className="col-10">User Name</div>
        <input
          className="col-10"
          type="text"
          name="username"
          value={username}
          onChange={handleChange}
        />
      </label>
      <label htmlFor="password" className="board center">
        <div className="col-10">Password</div>
        <input
          className="col-10"
          type="password"
          name="password"
          value={password}
          onChange={handleChange}
        />
      </label>
      <button type="submit">Submit</button>
    </form>
  );
};

ConnectedLogin.propTypes = {
  loginUser: PropTypes.func.isRequired,
  isLoggedIn: PropTypes.func.isRequired,
  history: PropTypes.func.isRequired,
};

const Login = connect(mapStateToProps, select)(ConnectedLogin);

export default Login;
