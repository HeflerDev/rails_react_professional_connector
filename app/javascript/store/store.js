import { createStore, combineReducers } from 'redux';
import userReducer from '../reducers/user';

const rootReducer = combineReducers({
  userReducer,
});

const store = createStore(rootReducer);

export default store;
