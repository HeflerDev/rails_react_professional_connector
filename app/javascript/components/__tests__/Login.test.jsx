import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../store/store';
import Login from '../Login';


describe('when rendering login', () => {
  test('it renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Login />
      </Provider>
    )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
