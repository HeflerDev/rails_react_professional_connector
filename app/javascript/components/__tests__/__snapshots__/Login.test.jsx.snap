// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when rendering login it renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    htmlFor="username"
  >
    <span>
      Username
    </span>
    <input
      name="username"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <label
    htmlFor="password"
  >
    <span>
      Password
    </span>
    <input
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </label>
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
